### PR TITLE
style: Sync tooling configuration from 4.0 branch

### PR DESCRIPTION
**Pull request by nbbeeken**
Update ESLint and bring in a modified version of our config from 4.0. ESLint now manages running and checking prettier formatting so switching branches should be a more pleasant experience. Also updating mocha and bringing in the config from the 4.0 branch provides a similar convenience.
Fixed a few linter errors most of which were replace `hasOwnProperty` direct calls with indirect `Object.prototype.hasOwnProperty.call`s.

### CODE REVIEW

**Review by reggi — APPROVED** [review]
LGTM, thanks for doing this. 🙇‍♂️

**Review by mbroadst — APPROVED** [review]
LGTM, with a nit about the eslint config.

**Review by emadum — APPROVED** [review]
LGTM! 👍

File: .eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  "root": true,
+  "parserOptions": {
+    "ecmaVersion": 2017
+  },
+  "plugins": [
+    "prettier"
+  ],
+  "extends": [
+    "eslint:recommended",
+    "prettier/@typescript-eslint",
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply .eslintrc.json]: Ah good catch!

**Pull request merged**